### PR TITLE
dynamic workflows

### PR DESCRIPTION
**Pull request by h0tw1r3**
* move images JSON out of workflows
* use workflow template to handle build and deploy
* ci, manual, and nightly workflows use template
* ci builds are dynamic based on changes to JSON or a dockerfile
* image push to registry is restricted to default branch

See it in "action":
* [Actions](https://github.com/h0tw1r3/litmusimage/actions)
* [Docker Hub registry](https://hub.docker.com/u/h0tw1r3)
* [Github Package Registry](https://github.com/h0tw1r3?ecosystem=container&tab=packages&tab=packages&ecosystem=container&q=litmusimage)

**Requires variables added to repository settings**
All are variables except DOCKER_PASSWORD, which is.a secret.

| name                | value   |
| ------------------- | ------ |
| `DOCKER_USERNAME`   | `waffleimage` |
| `DOCKER_PASSWORD` | `< copy secrets.DockerPassword >` |
| `DOCKER_REGISTRY`   | `docker.io` |
| `DOCKER_REPOSITORY` | `litmusimage` |

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -0,0 +1,75 @@
+name: CI/CD
+
+on:
```
Comment (jordanbreen28) [inline .github/workflows/ci.yml:3]: Could we update the triggers to match this for consistency?
https://github.com/puppetlabs/puppet_litmus/blob/05488913fe3dd0226793a3a0c7c8d233e02e18eb/.github/workflows/ci.yml#L3

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+      - name: Get changed files
+        id: changed-files
+        uses: tj-actions/changed-files@v40
+        with:
+          json: true
+          quotepath: false
+      - name: Create inputs
+        id: create-inputs
+        shell: python {0}
+        run: |
```
Comment (jordanbreen28) [inline .github/workflows/ci.yml:31]: I'm wondering if we should extract these - maybe into a `./scripts/` directory in the repo?

File: .github/workflows/template_build_deploy.yml [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+    steps:
+      - name: Checkout repository
+        uses: actions/checkout@v4
+        with:
+          fetch-depth: 1
+
+      - name: Create matrix from json
+        id: set-matrix
+        shell: python {0}
+        run: |
```
Comment (jordanbreen28) [inline .github/workflows/template_build_deploy.yml:46]: same as above

**Review by jordanbreen28 — COMMENTED** [review]
@h0tw1r3 sorry to have missed this! This is amazing work, I've left a couple of discussion points :)

**Comment by jordanbreen28** [comment]
required variables and secrets added to repo.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (h0tw1r3) [reply .github/workflows/ci.yml:3]: The design uses a template, so the `on` triggers are intended to be in each workflow file.

CI workflow automatically determines if images should be pushed to the docker repository based on the branch (main) using conditional steps in the template. Changing the on to include `workflow_dispatch` serves to purpose, and removing `push` would mean new images never get built / pushed until the nightly job.

Manual workflow includes the ability to set and pass inputs to the template, which the CI workflow does not (by design). Defaults are set to the same as CI.

File: .github/workflows/ci.yml (hunk 2 above)
Comment (h0tw1r3) [reply .github/workflows/ci.yml:31]: Yes, I was planning to do this but ran out of time.
Could this be pushed to a future improvement?
Specifically that would be
* External scripts
* Rewrite in nodejs
* Incorporate code scanning action

File: .github/workflows/template_build_deploy.yml (hunk 3 above)
Comment (h0tw1r3) [reply .github/workflows/template_build_deploy.yml:46]: see above comment, same response.

File: .github/workflows/ci.yml (hunk 2 above)
Comment (jordanbreen28) [reply .github/workflows/ci.yml:31]: Sure. Lets get a GH issue raised to track this.

**Review by jordanbreen28 — APPROVED** [review]

**Comment by jordanbreen28** [comment]
Great work @h0tw1r3 - thank you again!

**Pull request merged**